src/crimson/os/seastore/onode_manager/simple-fltree/onode_node.cc

```diff
@@ -0,0 +1,569 @@
+{
+  auto [prefix, suffix] = key_at(slot);
+  ghobject_t updated = oid;
+  prefix.update_oid(updated);
+  suffix.update_oid(updated);
+  return updated;
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+auto node_t<BlockSize, N, NodeType>::item_at(const key_t& key) const
+  -> const_item_t
+{
+  if constexpr (item_in_key) {
+    return key.child_addr;
+  } else {
+    assert(key.offset < BlockSize);
+    auto p = from_end(key.offset);
+    auto partial_key = reinterpret_cast<const partial_key_t*>(p);
+    p += size_of(*partial_key);
+    return *reinterpret_cast<const item_t*>(p);
+  }
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+void node_t<BlockSize, N, NodeType>::dump(std::ostream& os) const
+{
+  for (uint16_t i = 0; i < count; i++) {
+    const auto& [prefix, suffix] = key_at(i);
+    os << " [" << i << '/' << count - 1 << "]\n"
+       << "  key1 = (" << prefix << ")\n"
+       << "  key2 = (" << suffix << ")\n";
+    const auto& item = item_at(prefix);
+    if (_is_leaf()) {
+      os << " item = " << item << "\n";
+    } else {
+      os << " child = " << std::hex << item << std::dec << "\n";
+    }
+  }
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+char* node_t<BlockSize, N, NodeType>::from_end(uint16_t offset)
+{
+  auto end = reinterpret_cast<char*>(this) + BlockSize;
+  return end - static_cast<int>(offset);
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+const char* node_t<BlockSize, N, NodeType>::from_end(uint16_t offset) const
+{
+  auto end = reinterpret_cast<const char*>(this) + BlockSize;
+  return end - static_cast<int>(offset);
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+uint16_t node_t<BlockSize, N, NodeType>::used_space() const
+{
+  if constexpr (item_in_key) {
+    return count * sizeof(key_t);
+  } else {
+    if (count) {
+      return keys[count - 1].offset + count * sizeof(key_t);
+    } else {
+      return 0;
+    }
+  }
+}
+
+template<size_t BlockSize, int N, ntype_t NodeType>
+uint16_t node_t<BlockSize, N, NodeType>::capacity()
+{
+  auto p = reinterpret_cast<node_t*>(0);
+  return BlockSize - (reinterpret_cast<char*>(p->keys) -
+                      reinterpret_cast<char*>(p));
+}
+
+// TODO: if it's allowed to update 2 siblings at the same time, we can have
+//       B* tree
+template<size_t BlockSize, int N, ntype_t NodeType>
+constexpr uint16_t node_t<BlockSize, N, NodeType>::min_size()
+{
+  return capacity() / 2;
+}
+
+  // @return <minimum bytes to grab, maximum bytes to grab>
+template<size_t BlockSize, int N, ntype_t NodeType>
+constexpr std::pair<int16_t, int16_t>
+node_t<BlockSize, N, NodeType>::bytes_to_grab(uint16_t size)
+{
+  assert(size < min_size());
+  return {min_size() - size, capacity() - size};
+}
+
+// @return <minimum bytes to push, maximum bytes to push>
+template<size_t BlockSize, int N, ntype_t NodeType>
+constexpr std::pair<int16_t, int16_t> node_t<BlockSize, N, NodeType>::bytes_to_push(uint16_t size)
+{
```

Comment:
This doesn't appear to have a user yet, but it's not clear to me how "size" could be greater than capacity.  Is the intention that you project the size that will result from adding an element and pass that to bytes_to_push to determine how much to move to a sibling?

Reply:
yes. a typical use case is something like
```c++
  template<int UpN>
  read_ertr::future<update_t> remove_from(unsigned slot,
                                          InnerNode<BlockSize, UpN>* parent,
                                          unsigned whoami,
                                          Transaction& txn)
  {
    me->remove_from(slot);
    if (!parent) {
      // we can live with an empty root node
      return read_ertr::make_ready_future<update_t>();
    }
    if (count() == 0) {
      return read_ertr::make_ready_future<update_t>(update_t::make_removed(whoami));
    } else if (me->is_underflow(me->used_space())) {
      const auto [min_grab, max_grab] = me->bytes_to_add(me->used_space());
      // rebalance
      return grab_or_merge_left(*parent, whoami,
                                min_grab, max_grab, txn).safe_then(
        [parent, whoami, min_grab, max_grab, &txn, this](auto&& result) {
        if (result) {
          return read_ertr::make_ready_future<update_t>(result);
        } else {
          // yes, we leave this node underfull, because if we steal elements
          // from its siblings could lead to less optimum distribution of
          // indexed entries
          return grab_or_merge_right(*parent, whoami,
                                     min_grab, max_grab, txn);
        }
      });
    } else {
      return read_ertr::make_ready_future<update_t>();
    }
  }
```

please note, i already renamed the function name to "bytes_to_add()".